File: src/alerts/monitor.js
/**
 * Price Alert Monitor
 * Replaces TradingView webhooks — completely free.
 * Runs on a schedule, checks prices, triggers trades when conditions are met.
 */
import store from '../data/store.js';
import { fetchPrice } from '../data/market.js';
import * as portfolio from '../papertrading/portfolio.js';
import { calcStopLoss, calcTakeProfit } from '../strategy/risk.js';
import * as alpaca from '../brokers/alpaca.js';
import getConfig from '../config/index.js';

const config = getConfig();

/**
 * Load alert rules from data/alerts.json
 * Each rule:
 * {
 *   id: "btc_75k",
 *   symbol: "BTC",
 *   assetClass: "crypto",
 *   condition: "gte",         // gte, lte, cross_above, cross_below
 *   value: 75000,
 *   action: "buy",            // buy, sell
 *   message: "BTC crossed 75k",
 *   triggered: false          // Reset when condition clears
 * }
 */
function loadAlerts() {
  return store.read('alerts') || [];
}

function saveAlerts(alerts) {
  store.write('alerts', alerts);
}

/**
 * Check a single alert condition.
 */
function checkCondition(currentPrice, condition, threshold) {
  switch (condition) {
    case 'gte':       return currentPrice >= threshold;
    case 'lte':       return currentPrice <= threshold;
    case 'cross_above': return currentPrice >= threshold;
    case 'cross_below': return currentPrice <= threshold;
    default:          return false;
  }
}

/**
 * Run the alert monitor — checks all rules and executes triggered ones.
 */
export async function runAlertMonitor() {
  const alerts = loadAlerts();
  if (alerts.length === 0) {
    return { message: 'No alerts configured', triggered: [] };
  }

  const triggered = [];
  const defaults = config.strategy || {};

  for (const alert of alerts) {
    // Skip already-triggered alerts (wait for manual reset or condition reversal)
    if (alert.triggered && !alert.repeatable) continue;

    // Fetch current price
    const price = await fetchPrice(alert.symbol, alert.assetClass);
    if (!price) {
      console.log(`[Alerts] Could not fetch price for ${alert.symbol}`);
      continue;
    }

    // Check condition
    const met = checkCondition(price, alert.condition, alert.value);
    if (!met) {
      // If price moved away, reset triggered flag so it can fire again
      if (alert.triggered) {
        alert.triggered = false;
      }
      continue;
    }

    // Already triggered — skip unless repeatable
    if (alert.triggered) continue;

    // Condition met! Execute the trade on paper trading engine
    const port = portfolio.loadPortfolio(config.paperBalance);
    const balance = port.balances?.USD || 0;
    const riskPerTrade = alert.riskPerTrade || defaults.riskPerTrade || 0.015;
    const positionValue = balance * riskPerTrade * 10;
    const quantity = positionValue / price;

    const stopLossPct = alert.stopLossPct || defaults.stopLossPct || 0.02;
    const riskRewardRatio = alert.riskRewardRatio || defaults.riskRewardRatio || 2;
    const stopLoss = calcStopLoss(price, 'long', stopLossPct);
    const takeProfit = calcTakeProfit(price, 'long', riskRewardRatio * stopLossPct);

    // Update portfolios
    portfolio.updatePositionPrices(port, { [alert.symbol]: price });

    let execution;
    if (alert.action === 'buy' && !portfolio.hasOpenPosition(port, alert.symbol)) {
      const pos = portfolio.openPosition(port, {
        symbol: alert.symbol,
        assetClass: alert.assetClass || 'crypto',
        side: 'long',
        quantity,
        entryPrice: price,
        stopLoss,
        takeProfit,
        reason: `Alert: ${alert.message || alert.symbol} ${alert.condition} ${alert.value}`,
      });
      execution = { type: 'entry', symbol: alert.symbol, price, quantity, positionId: pos.id };
    } else if (alert.action === 'sell') {
      const existingPos = port.positions.find(p => p.symbol === alert.symbol && p.status === 'open');
      if (existingPos) {
        const closed = portfolio.closePosition(port, existingPos.id, price, 'alert_signal');
        execution = { type: 'exit', symbol: alert.symbol, price, pnl: closed?.pnl };
      }
    }

    if (execution) {
      alert.triggered = true;
      alert.lastTriggered = new Date().toISOString();
      alert.triggerPrice = price;
      triggered.push({ alert: alert.id, ...execution });

      // Also place order on Alpaca if connected
      if (alpaca.isConnected()) {
        try {
          const alpacaOrder = await alpaca.placeOrder({
            symbol: alert.symbol,
            assetClass: alert.assetClass || 'crypto',
            qty: quantity.toFixed(8),
            side: alert.action,
            type: 'market',
            stopLoss: stopLoss,
            takeProfit: takeProfit,
          });
          execution.alpacaOrderId = alpacaOrder.id;
          execution.alpacaStatus = alpacaOrder.status;
        } catch (alpacaErr) {
          console.error(`[Alerts] Alpaca order failed: ${alpacaErr.message}`);
          execution.alpacaError = alpacaErr.message;
        }
      }
    }
  }

  saveAlerts(alerts);
  return { message: 'Monitor complete', triggered, alertsChecked: alerts.length };
}

/**
 * Add a new alert rule.
 */
export function addAlert(rule) {
  const alerts = loadAlerts();
  const id = rule.id || `${rule.symbol.toLowerCase()}_${Date.now()}`;
  alerts.push({
    id,
    symbol: rule.symbol.toUpperCase(),
    assetClass: rule.assetClass || 'crypto',
    condition: rule.condition || 'gte',
    value: rule.value,
    action: rule.action || 'buy',
    message: rule.message || '',
    riskPerTrade: rule.riskPerTrade,
    stopLossPct: rule.stopLossPct,
    riskRewardRatio: rule.riskRewardRatio,
    repeatable: rule.repeatable || false,
    triggered: false,
    createdAt: new Date().toISOString(),
  });
  saveAlerts(alerts);
  return { id, message: `Alert '${id}' added` };
}

/**
 * Remove an alert rule.
 */
export function removeAlert(id) {
  const alerts = loadAlerts();
  const idx = alerts.findIndex(a => a.id === id);
  if (idx === -1) return { error: 'Alert not found' };
  alerts.splice(idx, 1);
  saveAlerts(alerts);
  return { message: `Alert '${id}' removed` };
}

export default { runAlertMonitor, addAlert, removeAlert };

File: src/alerts/monitor.py
"""Alerts engine — condition-based price alerts.

Ported from hermes-trading-agent (HTA). Persists alert rules in
``data/alerts.json`` and evaluates them on a schedule (every
``AUTOMATION_INTERVAL_MS`` ms by default).

Alert shape:
    {
        "id": "alert_abc123",
        "symbol": "BTC",
        "assetClass": "crypto" | "stock" | "forex",
        "condition": "gte" | "lte" | "cross_above" | "cross_below",
        "value": 75000.0,
        "action": "buy" | "sell" | "alert",
        "message": "...",
        "repeatable": false,
        "triggered": false,
        "lastTriggeredAt": null
    }
"""

import logging
import threading
from datetime import datetime
from typing import Any, Dict, List, Optional

from src.alerts.store import AlertStore

logger = logging.getLogger(__name__)


# Track last seen price per (alert_id) to detect crossings
_PRICE_HISTORY: Dict[str, float] = {}
_HISTORY_LOCK = threading.Lock()


# ── Condition evaluation ────────────────────────────────────


def _check_condition(
    current: float, condition: str, threshold: float, previous: Optional[float]
) -> bool:
    if condition == "gte":
        return current >= threshold
    if condition == "lte":
        return current <= threshold
    if condition == "cross_above":
        return previous is not None and previous < threshold <= current
    if condition == "cross_below":
        return previous is not None and previous > threshold >= current
    return False


# ── Price fetching ──────────────────────────────────────────


def _fetch_price(symbol: str, asset_class: str) -> Optional[float]:
    """Try multiple sources in priority order: Binance → CoinGecko → Yahoo."""
    asset_class = (asset_class or "crypto").lower()

    # Try Binance for crypto
    if asset_class == "crypto":
        try:
            from src.data.sources.binance import BinanceSource

            p = BinanceSource().fetch_price(symbol)
            if p:
                return p
        except Exception as e:
            logger.debug(f"Binance fetch failed for {symbol}: {e}")

        # Fallback: CoinGecko
        try:
            from src.data.sources.coingecko import CoinGeckoSource

            src = CoinGeckoSource()
            md = src.get_market_data(symbol)
            if md and md.get("current_price"):
                return float(md["current_price"])
        except Exception as e:
            logger.debug(f"CoinGecko fetch failed for {symbol}: {e}")

    # Yahoo Finance fallback (works for stocks, ETFs, and some crypto pairs)
    try:
        import yfinance as yf

        ticker = yf.Ticker(symbol)
        hist = ticker.history(period="1d")
        if not hist.empty:
            return float(hist["Close"].iloc[-1])
    except Exception as e:
        logger.debug(f"Yahoo fetch failed for {symbol}: {e}")

    return None


# ── Public API ──────────────────────────────────────────────


def add_alert(
    symbol: str,
    asset_class: str,
    condition: str,
    value: float,
    action: str = "alert",
    message: str = "",
    repeatable: bool = False,
) -> Dict[str, Any]:
    """Add a new alert rule."""
    store = AlertStore()
    return store.add(
        symbol=symbol,
        asset_class=asset_class,
        condition=condition,
        value=value,
        action=action,
        message=message,
        repeatable=repeatable,
    )


def list_alerts() -> List[Dict[str, Any]]:
    return AlertStore().list()


def remove_alert(alert_id: str) -> bool:
    ok = AlertStore().remove(alert_id)
    with _HISTORY_LOCK:
        _PRICE_HISTORY.pop(alert_id, None)
    return ok


def reset_alerts() -> int:
    return AlertStore().reset_all()


def run_monitor() -> Dict[str, Any]:
    """Evaluate every alert; execute actions on triggered ones.

    Returns a summary dict with counts and triggered alert list.
    """
    store = AlertStore()
    alerts = store.list()
    triggered: List[Dict[str, Any]] = []

    for alert in alerts:
        # Skip one-shot alerts already triggered (unless repeatable)
        if alert.get("triggered") and not alert.get("repeatable"):
            continue

        price = _fetch_price(alert["symbol"], alert.get("assetClass", "crypto"))
        if price is None:
            continue

        with _HISTORY_LOCK:
            previous = _PRICE_HISTORY.get(alert["id"])
            _PRICE_HISTORY[alert["id"]] = price

        if _check_condition(price, alert["condition"], alert["value"], previous):
            result = _execute_action(alert, price)
            triggered.append(
                {
                    "id": alert["id"],
                    "symbol": alert["symbol"],
                    "condition": alert["condition"],
                    "value": alert["value"],
                    "currentPrice": price,
                    "action": alert["action"],
                    "executed": result,
                }
            )
            store.mark_triggered(alert["id"])

    return {
        "checked": len(alerts),
        "triggered": triggered,
        "timestamp": datetime.utcnow().isoformat() + "Z",
    }


def _execute_action(alert: Dict[str, Any], price: float) -> Dict[str, Any]:
    """Execute the alert's action. For 'buy'/'sell' we route to PaperTrader."""
    action = alert.get("action", "alert")
    if action not in ("buy", "sell"):
        return {"executed": False, "reason": f"action={action} (notify-only)"}

    try:
        from src.execution.paper import PaperTrader

        trader = PaperTrader()
        result = trader.execute_signal(
            {
                "symbol": alert["symbol"],
                "direction": "BUY" if action == "buy" else "SELL",
                "confidence": 0.7,
                "reasoning": f"Alert {alert['id']}: {alert['message'] or alert['condition'] + ' ' + str(alert['value'])} @ ${price:.4f}",
            }
        )
        return {"executed": bool(result), "price": price}
    except Exception as e:
        logger.error(f"Alert action execution failed: {e}")
        return {"executed": False, "error": str(e)}
